### PR TITLE
Introduce `DirectReturn` check

### PR DESCRIPTION
**Pull request by Stephan202**
This PR replaces #1. I decided to pick this task up because the topic was raised after @rickie's recent Amsterdam JUG presentation. CC @SimonBaars.

Suggested commit message:
```
Introduce `DirectReturn` check (#513)
```

### CODE REVIEW

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/DirectReturn.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+ * statement.
+ */
+@AutoService(BugChecker.class)
+@BugPattern(
+    summary = "Variable assignment is redundant; value can be returned directly",
+    link = BUG_PATTERNS_BASE_URL + "DirectReturn",
+    linkType = CUSTOM,
+    severity = SUGGESTION,
+    tags = SIMPLIFICATION)
+public final class DirectReturn extends BugChecker implements BlockTreeMatcher {
```
Comment (Stephan202) [inline error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/DirectReturn.java:43]: Naming-wise, "something something unnecessary assignment" may be nicer. But (a) we're not performing a flow analysis that flags _all_ unnecessary variable assignments and (b) a name that accurately describes the subset covered would be quite unwieldy. Unless I'm just not creative enough, of course :upside_down_face:. Suggestions welcome.

**Comment by Stephan202** [comment]
Just realized: ideally we do not suggest inlining if (a) the variable assigned to is of a different type than the method's return type and (b) this difference may influence the return value. Prime example here is Mockito's nullary `mock()` method (see also #454). I'll come back to this PR to see how we can best cover this case.

**Comment by Stephan202** [comment]
Rebased and added a commit to cover this case. Did make it Mockito-specifc, because attempts at generalization became too contrived.

**Review by Stephan202 — COMMENTED** [review]
Added one more commit.

**Review by rickie — COMMENTED** [review]
Flushing one comment I had. Will continue later today.

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/DirectReturn.java [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+  }
+
+  /**
+   * Tells whether inlining the given expression to the associated return statement can be done
+   * safely.
+   *
+   * <p>Inlining is generally safe, but in rare cases the operation may have a functional impact.
+   * The sole case considered here is the inlining of a Mockito mock or spy construction without an
+   * explicit type. In such a case the type created depends on context, such as the method's return
+   * type.
```
Comment (werli) [inline error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/DirectReturn.java:122]: Great to have added this exclusion. Are we sure the exclusion is exhaustive however? My guess is: probably never _fully_, but this is a best-effort solution that should handle the majority of EPS users.

**Review by werli — APPROVED** [review]
The code itself and the added test cases were rather straightforward, nice work @Stephan202 :+1:

One more point about the exhaustiveness of the mockito exclusion, but I wouldn't expect more changes here as part of this PR. Hence I'm approving.

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/DirectReturn.java (hunk 2 above)
Comment (Stephan202) [reply error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/DirectReturn.java:122]: Nope, this is not exhaustive, unfortunately. I did play with some more generic code, but it became way more complex than I could justify for this feature. So in the end gave up on that idea.

**Review by Stephan202 — COMMENTED** [review]
Tnx for the reviews!

File: error-prone-contrib/src/test/java/tech/picnic/errorprone/bugpatterns/DirectReturnTest.java [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+            "    return variable;",
+            "  }",
+            "",
+            "  Object redundantSpyWithExplicitTypeVariableDeclaration() {",
+            "    // BUG: Diagnostic contains:",
+            "    String variable = spy(String.class);",
+            "    return variable;",
+            "  }",
+            "",
+            "  Object salienSpyTypeVariableDeclaration() {",
```
Comment (rickie) [inline error-prone-contrib/src/test/java/tech/picnic/errorprone/bugpatterns/DirectReturnTest.java]:
```suggestion
            "  Object salientSpyTypeVariableDeclaration() {",
```

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/util/MoreASTHelpers.java [hunk 4]
```diff
@@ -46,4 +50,33 @@ public static ImmutableList<MethodTree> findMethods(CharSequence methodName, Vis
   public static boolean methodExistsInEnclosingClass(CharSequence methodName, VisitorState state) {
     return !findMethods(methodName, state).isEmpty();
   }
+
+  /**
+   * Returns the {@link MethodTree} from which control flow would exit if there would be a {@code
```
Comment (rickie) [inline error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/util/MoreASTHelpers.java]:
```suggestion
   * Returns the {@link MethodTree} from which the control flow would exit if there would be a {@code
```
Sounds a bit better to me 🤔.

**Review by rickie — APPROVED** [review]
Really cool check 🚀.

Have some small comments and an extra test case or two. Let me know what you think 😄.

**Comment by rickie** [comment]
Aii, misclicked and merged instead of rebased 😅.

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/util/MoreASTHelpers.java [hunk 5]
```diff
@@ -46,4 +50,33 @@ public static ImmutableList<MethodTree> findMethods(CharSequence methodName, Vis
   public static boolean methodExistsInEnclosingClass(CharSequence methodName, VisitorState state) {
     return !findMethods(methodName, state).isEmpty();
   }
+
+  /**
+   * Returns the {@link MethodTree} from which the control flow would exit if there would be a
```
Comment (Stephan202) [inline error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/util/MoreASTHelpers.java:55]: I think we should omit "the" here. (Cause this is not about a "definite" control flow.)

File: error-prone-contrib/src/test/java/tech/picnic/errorprone/bugpatterns/DirectReturnTest.java [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+            "    if (true) {",
+            "      return variable;",
+            "    } else {",
+            "      // BUG: Diagnostic contains:",
+            "      variable = \"foo\";",
+            "      return variable;",
+            "    }",
+            "  }",
+            "",
+            "  Supplier<String> redundantAssignmentInLambda() {",
```
Comment (Stephan202) [inline error-prone-contrib/src/test/java/tech/picnic/errorprone/bugpatterns/DirectReturnTest.java:132]:
```suggestion
            "  Supplier<String> redundantAssignmentInsideLambda() {",
```
Also matches the method above.

**Review by Stephan202 — COMMENTED** [review]
Tnx; changes LGTM otherwise

**Comment by rickie** [comment]
Applied and rebased. Will merge once 🟢, thanks for taking a look so quickly.

**Comment by Stephan202** [comment]
Let's gooooo :rocket:

**Pull request merged**